### ISSUE TITLE
tf.nn.sparse_softmax_cross_entropy_with_logits loses a finite first-derivative signal for float64 input

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#126636]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13

### Current behavior?

`tf.nn.sparse_softmax_cross_entropy_with_logits` returns zero for one logit-gradient component whose correctly rounded `float64` reference value is finite, normal, and nonzero.

Sparse softmax cross entropy is invariant to adding the same constant to all logits, so the gradient with respect to logits should sum to zero. In the reproducer below, the expected gradient is approximately `[-5.551115123125776e-17, 5.551115123125776e-17]`.

However, TensorFlow returns `[[0.0, 5.551115123125776e-17]]`. The first gradient component is incorrectly collapsed to zero, and the resulting logit gradient no longer preserves the expected zero-sum relation.

Expected behavior?

The first-order autodiff result for `tf.nn.sparse_softmax_cross_entropy_with_logits` should preserve the finite, normal, nonzero `float64` gradient component. For the given logits and sparse label, the gradient should be close to `[[-5.551115123125776e-17, 5.551115123125776e-17]]`, and the gradient sum should be zero.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

dtype = tf.float64

expected = [-5.551115123125776e-17, 5.551115123125776e-17]

logits = tf.Variable([[37.42994775023705, 0.0]], dtype=dtype)
labels = tf.constant([0], dtype=tf.int32)

with tf.GradientTape() as tape:
    loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
        labels=labels,
        logits=logits,
    )
    objective = tf.reduce_sum(loss)

actual = tape.gradient(objective, logits)
grad_sum = tf.reduce_sum(actual)

print("loss:")
print(loss.numpy().tolist())
print("actual:")
print(actual.numpy().tolist())
print("expected:")
print(expected)
print("gradient sum:")
print(float(grad_sum))

if actual.numpy()[0, 0] == 0.0 and expected[0] != 0.0:
    print("BUG REPRODUCED: tf.nn.sparse_softmax_cross_entropy_with_logits loses a finite first-derivative signal")
else:
    print("not reproduced")
```

### Relevant log output

```shell
loss:
[0.0]
actual:
[[0.0, 5.551115123125776e-17]]
expected:
[-5.551115123125776e-17, 5.551115123125776e-17]
gradient sum:
5.551115123125776e-17
BUG REPRODUCED: tf.nn.sparse_softmax_cross_entropy_with_logits loses a finite first-derivative signal
```

**Comment by Kayyuri** [comment]
Hi @ALinrunrun, Thank you for reporting this issue and providing such a clear, detailed reproduction case! It looks like there's an unexpected behavior in `tf.nn.sparse_softmax_cross_entropy_with_logits` when calculating gradients with `float64` inputs, specifically losing a finite first-derivative signal. We appreciate you taking the time to provide the standalone code and the expected vs. actual output, which makes it very easy to understand and reproduce. We will investigate this further.

**Comment by Kayyuri** [comment]
Hi @ALinrunrun,
A fix for this issue has been proposed in this [PR](https://github.com/tensorflow/tensorflow/pull/126804). Once it is merged, this issue will be resolved.

Hi @GodlyDonuts,
Thank you for raising the PR to address this issue. We truly appreciate your initiative and the effort in contributing the fix.
